### ISSUE TITLE
Sandboxing Python with Win32 App Isolation

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#10997]
Sandboxing Python with Win32 App Isolation<br>
<br>
https://ift.tt/Ptkd3rb<br>
<br>
Tian Gao<br>
<br>
<div>
<section data-js="panel-article-header" data-extra-context>
<div>
<div>
<div>
<span>
<span>
Written By
</span>
</span>
<span>
<span>
published
</span>
<span>
March 6, 2024
</span>
</span>
</div>
</div>
</div></section><div data-modular-content-collection>
<section data-js="panel" data-type="wysiwyg-with-aside" data-index="0" data-modular-content>
<div>
<div>
<div>
<p><span data-contrast="auto">Sandboxing Python can be very useful for developers but has been challenging due to the flexibility of CPython implementation. It is particularly useful in many scenarios to have sandboxed Python.  For example, on any website that needs to execute arbitrary code from users, like coding practice sites or online interpreters. It could also be useful to help prevent increasing attacks on LLMs (Large Language Models).</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<p><span data-contrast="auto">Win32 App Isolation (https://ift.tt/2qis6OY) lights up a different path to sandboxing Python – not on a Python library level, but on an application/OS level.</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<p><span data-contrast="auto">Win32 App Isolation is designed to isolate Win32 application by creating a security boundary between the application and the OS, using the combination of AppContainer, virtualized resources and brokering file system, which helps to prevent the application from compromising the operating system.</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<p><span data-contrast="auto">Following are the specifics requirements and guidance for sandboxing</span> <span data-contrast="auto">CPython with Win32 App Isolation on recent Windows Insider releases. It takes about 30-60 minutes and anyone with the insider version of Windows can do it on their local machine.</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<p><span data-contrast="auto">To use Win32 App Isolation, you need the insider version of Windows with version number &gt; 25357, and the MSIX Packaging Tool from https://ift.tt/DRm3lCO</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<p><span data-contrast="auto">Step One:</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<p><span data-contrast="auto">Download the Python installer for Windows from python.org. 3.12.2 was used for this specific example.</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<p><span data-contrast="auto">Step Two:</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<p><span data-contrast="auto">Package it by following the documentation (</span><a href="https://github.com/microsoft/win32-app-isolation/blob/main/docs/packaging/msix-packaging-tool.md"><span data-contrast="none">https://github.com/microsoft/win32-app-isolation/blob/main/docs/packaging/msix-packaging-tool.md</span></a><span data-contrast="auto">). You may need a certificate to sign the package to install it (https://ift.tt/hpa307H).</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<p><span data-contrast="auto">Add <p><span><span>isolatedWin32-PromptForAccess </span></span></p>
<code data-enlighter-language="generic">isolatedWin32-PromptForAccess </code></span><span data-contrast="auto">capability for demonstration purposes, so you can allow it to access certain resources in prompts. Without the capability, all access to the user files that are not explicitly set to be available to the application would be denied.</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<p><span data-contrast="auto">Add an alias <p><span><span>python3.</span><span>12</span><span>.</span><span>exe</span><span> </span></span></p>
<code data-enlighter-language="generic">python3.12.exe </code></span><span data-contrast="auto">to use it from PowerShell.</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<p><span data-contrast="auto">Step Three:</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<p><span data-contrast="auto">Install the package. Use the defined alias <p><span><span>python3.</span><span>12</span><span>.</span><span>exe</span></span></p>
<code data-enlighter-language="generic">python3.12.exe</code> </span><span data-contrast="auto">in PowerShell to bring up the Python interpreter.</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<figure class="wp-image wp-image--no-caption"><div><img src="https://blogs.windows.com/wp-content/uploads/prod/sites/3/2024/02/1.png" alt="A screenshot of a computer" width="624" class="wp-image-56928" height="355"></div></figure>
<p><span data-contrast="auto">When starting the Python interpreter, it will prompt you to ask permission to the current working directory. Python itself would be brought up with either option – but this determines whether Python has access to the files in the current directory.</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<p><span data-contrast="auto">For example, if you clicked “No” here, you would not be able to list the files in the current directory anymore:</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<figure class="wp-image wp-image--no-caption"><div><img src="https://blogs.windows.com/wp-content/uploads/prod/sites/3/2024/02/Picture2.png" alt="A screenshot of a computer" width="624" class="wp-image-56930" height="356"></div></figure>
<p><span data-contrast="auto">The selection will be remembered so the application will only prompt once for any file/folder. The user can reset the file permissions in Settings – Privacy &amp; security – File system:</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<figure class="wp-image wp-image--no-caption"><div><img src="https://blogs.windows.com/wp-content/uploads/prod/sites/3/2024/02/Picture3.png" alt="A screenshot of a computer" width="624" class="wp-image-56931" height="466"></div></figure>
<p><span data-contrast="auto">The Python interpreter should work well on its own given no code has been changed. </span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<figure class="wp-image wp-image--no-caption"><div><img src="https://blogs.windows.com/wp-content/uploads/prod/sites/3/2024/02/Picture4.png" alt="A screenshot of a computer" width="624" class="wp-image-56932" height="356"></div></figure>
<p><span data-contrast="auto">The more interesting aspects to check out are the “sandboxing” features – network and file system.</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<p><span data-contrast="auto">But before that, try to execute a script using isolated Python.</span><span data-contrast="auto"> </span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<figure class="wp-image wp-image--no-caption"><div><img src="https://blogs.windows.com/wp-content/uploads/prod/sites/3/2024/02/Picture5.png" alt="A screenshot of a computer" width="624" class="wp-image-56933" height="357"></div></figure>
<p><span data-contrast="auto">As you are trying to access and execute a user file (the script), it will prompt for the access. If denied, the script will not run:</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<figure class="wp-image wp-image--no-caption"><div><img src="https://blogs.windows.com/wp-content/uploads/prod/sites/3/2024/02/Picture6.png" alt="A screenshot of a computer" width="624" class="wp-image-56934" height="357"></div></figure>
<p><span data-contrast="auto">Allow it to review some examples, starting with network access.</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<p><span data-contrast="auto">As you did not provide any network capabilities when packaging, Python will not have any access to the network. Confirm this by running this script:</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<pre data-enlighter-language="generic">def test_urllib(): 
    with urllib.request.urlopen("http://www.microsoft.com") as response: 
        print(response.read())</pre>
<figure class="wp-image wp-image--no-caption"><div><img src="https://blogs.windows.com/wp-content/uploads/prod/sites/3/2024/02/Picture7.png" alt="A screenshot of a computer" width="624" class="wp-image-56935" height="355"></div></figure>
<p><span data-contrast="auto">Also try to access the file system.</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<pre data-enlighter-language="generic">def test_open(): 
    with open("../data/test.txt", "w") as f: 
        f.write("test")</pre>
<figure class="wp-image wp-image--no-caption"><div><img src="https://blogs.windows.com/wp-content/uploads/prod/sites/3/2024/02/Picture8.png" alt="A screenshot of a computer" width="624" class="wp-image-56936" height="354"></div></figure>
<p><span data-contrast="auto">A prompt was shown to request the access to the file to be written, and you can deny that access by clicking “No”</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<figure class="wp-image wp-image--no-caption"><div><img src="https://blogs.windows.com/wp-content/uploads/prod/sites/3/2024/02/Picture10.png" alt="A screenshot of a computer" width="624" class="wp-image-56937" height="354"></div></figure>
<p><span data-contrast="auto">It will not work with the native APIs either:</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<pre data-enlighter-language="generic">def test_native(): 
    handle = ctypes.windll.kernel32.CreateFileW("../data/test.txt", 0x10000000, 0, None, 1, 0x80, 0) 
    if handle &gt; 0: 
        ctypes.windll.kernel32.CloseHandle(handle) 
    else: 
        raise RuntimeError("Failed to get the handle")</pre>
<figure class="wp-image wp-image--no-caption"><div><img src="https://blogs.windows.com/wp-content/uploads/prod/sites/3/2024/02/Picture11.png" alt="A screenshot of a computer" width="624" class="wp-image-56938" height="356"></div></figure>
<p><span data-contrast="auto">You can also test <p><span><span>os.</span><span>system</span><span>()</span></span></p>
<code data-enlighter-language="generic">os.system()</code></span><span> </span><span data-contrast="auto">and <p><span><span>subprocess</span></span></p>
<code data-enlighter-language="generic">subprocess</code></span><span data-contrast="auto">:</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<pre data-enlighter-language="generic">def test_os_system(): 
    os.system("powershell -command cat c:/data/test/bin/data/test.txt") 
def test_subprocess(): 
    subprocess.run(["powershell", "-command", "cat", "c:/data/test/bin/data/test.txt"], shell=True)</pre>
<figure class="wp-image wp-image--no-caption"><div><img src="https://blogs.windows.com/wp-content/uploads/prod/sites/3/2024/02/Picture12.png" alt="A screenshot of a computer" width="624" class="wp-image-56939" height="355"></div></figure>
<p><span data-contrast="auto">Importing modules by modifying <p><span><span>sys.</span><span>path</span><span> </span></span></p>
<code data-enlighter-language="generic">sys.path </code></span><span data-contrast="auto">will trigger the prompt too:</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<pre data-enlighter-language="generic">def test_import(): 
    sys.path.append("../data") 
    import example</pre>
<figure class="wp-image wp-image--no-caption"><div><img src="https://blogs.windows.com/wp-content/uploads/prod/sites/3/2024/02/Picture13.png" alt="A screenshot of a computer" width="624" class="wp-image-56940" height="354"></div></figure>
<p><span data-contrast="auto">Clicking “no” will deny the access to the module (<p><span><span>example.</span><span>py</span></span></p>
<code data-enlighter-language="generic">example.py</code></span><span data-contrast="auto">is in the data directory and clicking “yes” would successfully import it):</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<figure class="wp-image wp-image--no-caption"><div><img src="https://blogs.windows.com/wp-content/uploads/prod/sites/3/2024/02/Picture14.png" alt="A screenshot of a computer" width="624" class="wp-image-56941" height="354"></div></figure>
<p><span data-contrast="auto">From the examples above, by introducing Win32 App Isolation you can help prevent Python from accessing the file system and the network without consent.</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<p><span data-contrast="auto">There are two obvious questions:</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
<ol>
<li>How can you “click” the prompts if you want to deploy this on your server? As mentioned above, <p><span><span>isolatedWin32-promptForAccess</span></span></p>
<code data-enlighter-language="generic">isolatedWin32-promptForAccess</code>is only added for demonstration purposes, and it is not needed for isolation. Without it, your system would simply deny any access to user files, thus helping to prevent attacks like ransomware.</li>
<li>Then how can you grant access to the application for required files like modules, scripts, and data? There are several ways to do it:<span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span>
<ul>
<li>You can package the modules you need into your app, then they will be accessible by default.</li>
<li>You can put the file you need access to into the app’s profile in <p><span><span>%localappdata%</span></span></p>
<code data-enlighter-language="generic">%localappdata%</code> where the app is granted access automatically.</li>
<li>You can put the file to a “publisher directory”, which is any directory with a name ending with your app’s publisher ID in <p><span><span>c:\ProgramData</span></span></p>
<code data-enlighter-language="generic">c:\ProgramData</code>. (<a href="https://github.com/microsoft/win32-app-isolation/blob/main/docs/fundamentals/consent.md"><span data-contrast="none">win32-app-isolation/docs/fundamentals/consent.md at main · microsoft/win32-app-isolation (github.com)</span></a><span data-contrast="auto">)</span>
</li>
<li>You can set the access of files and directories with icalcs to allow your app or all apps to access them.</li>
</ul>
</li>
</ol>
<p><span data-contrast="auto">With the technology, the users can create much lighter solutions when they try to execute code from an untrusted source. Isolated Python can live in parallel with normal Python and perform as a low-privilege interpreter on the side. Or, with correct configuration, it can be used on its own which could help protect the system from being compromised while keeping all the functionalities from Python.</span><span data-ccp-props='{"201341983":0,"335559739":160,"335559740":279}'> </span></p>
</div>
</div>
</div>
</section>
</div>
</div><br>
<br>
<br>
<br>
<br>
<br>
via Windows Developer Blog https://ift.tt/U39tkDE<br>
<br>
April 2, 2024 at 07:19PM
